tests/test_init.py

```diff
@@ -53,3 +60,103 @@ def test_parse_bad_patchwork_url(self):
+        self.m_io_open = Mock()
+        self.m_io_open.__enter__ = lambda *args: self.m_io_open
+        self.m_io_open.__exit__ = lambda *args: None
+        self.ctx_io_open = mock.patch('io.open',
+                                      Mock(return_value=self.m_io_open))
+        self.kernel_tarball = 'linux-4.16.0.tar.gz'
+        self.success_str = 'Tarball successfully created in ./{}\n'
+        self.success_str = self.success_str.format(self.kernel_tarball)
+
+    def tearDown(self):
+        shutil.rmtree(self.tmpdir)
+
+    def test_mktgz_clean(self):
+        """
+        Check if `mrproper` target is called when `clean` is True and not
+        called when `clean` is False.
+        """
+        with self.ctx_popen, self.ctx_check_call as m_check_call:
+            self.assertRaises(Exception, self.kbuilder.mktgz, clean=True)
+            self.assertEqual(
+                m_check_call.mock_calls[0],
+                mock.call(['make', '-C', self.tmpdir, 'mrproper'])
+            )
+            m_check_call.reset_mock()
+            self.assertRaises(Exception, self.kbuilder.mktgz, clean=False)
+            self.assertNotEqual(
+                m_check_call.mock_calls[0],
+                mock.call(['make', '-C', self.tmpdir, 'mrproper']),
+            )
+
+    def test_mktgz_timeout(self):
+        """
+        Check if timeout error is raised when kernel building takes longer than
+        specified timeout.
+        """
+        self.m_popen.poll = Mock(side_effect=[None, None, -15])
+        self.m_popen.returncode = -15
+        from time import sleep as real_sleep
+
+        def m_sleep(seconds):
+            """sleep function but acelerated 100x"""
+            real_sleep(seconds/100)
+        ctx_sleep = mock.patch('time.sleep', m_sleep)
+        with ctx_sleep, self.ctx_check_call, self.ctx_popen:
+            self.assertRaises(skt.CommandTimeoutError,
+                              self.kbuilder.mktgz, timeout=0.001)
+
+    def test_mktgz_parsing_error(self):
+        """Check if ParsingError is raised when no kernel is found in stdout"""
+        self.m_io_open.readlines = Mock(return_value=['foo\n', 'bar\n'])
+        with self.ctx_popen, self.ctx_check_call, self.ctx_io_open:
+            self.assertRaises(skt.ParsingError, self.kbuilder.mktgz)
+
+    def test_mktgz_ioerror(self):
+        """Check if IOError is raised when tarball path does not exist"""
+        self.m_io_open.readlines = Mock(
+            return_value=['foo\n', self.success_str]
+        )
+        with self.ctx_io_open, self.ctx_popen, self.ctx_check_call:
+            self.assertRaises(IOError, self.kbuilder.mktgz)
+
+    def test_mktgz_make_fail(self):
+        """
+        Check if subprocess.CalledProcessError is raised when make command
+        fails to spawn.
+        """
+        self.m_io_open.readlines = Mock(
+            return_value=['foo\n', self.success_str, 'bar']
+        )
```

Comment:
Hmm, we shouldn't get to actually reading the "make" output if it fails to spawn, therefore this is unnecessary and is confusing. Can we remove this, and the `ctx_io_open` use? Also, `ctx_check_call` shouldn't be used either, right?

Reply:
Yeah! Done!